File: infra/communication/go/dds/bus.go
package dds

import (
	"context"
	"fmt"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"

	communication "github.com/pacific-rim/pacific-rim/infra/communication/go/contracts"
	"github.com/pacific-rim/pacific-rim/infra/communication/go/core"
)

type ByteHandler func(context.Context, []byte) error

type ByteClient interface {
	Connect(context.Context, CycloneDDSConfig) error
	Close(context.Context) error
	PreparePublish(context.Context, TopicConfig) error
	Publish(context.Context, TopicConfig, []byte) error
	Subscribe(context.Context, Subscription, ByteHandler) error
}

type ByteClientFactory func(CycloneDDSConfig) (ByteClient, error)

type Bus struct {
	config         CycloneDDSConfig
	typeName       string
	qos            map[string]string
	client         ByteClient
	adapters       map[string]RPCAdapter
	kind           communication.TransportKind
	publishReadyMu sync.Mutex
	publishReady   map[string]bool
}

func NewBus(config CycloneDDSConfig, typeName string, qos map[string]string, client ByteClient) *Bus {
	return NewBusWithRPCAdapters(config, typeName, qos, client, nil)
}

func NewBusWithRPCAdapters(
	config CycloneDDSConfig,
	typeName string,
	qos map[string]string,
	client ByteClient,
	adapters map[string]RPCAdapter,
) *Bus {
	return NewBusWithKindAndRPCAdapters(communication.TransportCycloneDDS, config, typeName, qos, client, adapters)
}

func NewBusWithKindAndRPCAdapters(
	kind communication.TransportKind,
	config CycloneDDSConfig,
	typeName string,
	qos map[string]string,
	client ByteClient,
	adapters map[string]RPCAdapter,
) *Bus {
	return &Bus{
		config:       config,
		typeName:     typeName,
		qos:          qos,
		client:       client,
		adapters:     normalizeAdapters(adapters),
		kind:         kind,
		publishReady: map[string]bool{},
	}
}

func NewBusFactory(factory ByteClientFactory) core.Factory {
	return NewBusFactoryWithRPCAdapters(factory, nil)
}

func NewBusFactoryWithRPCAdapters(factory ByteClientFactory, adapters map[string]RPCAdapter) core.Factory {
	return NewBusFactoryWithKindAndRPCAdapters(communication.TransportCycloneDDS, factory, adapters)
}

func NewBusFactoryWithKindAndRPCAdapters(
	kind communication.TransportKind,
	factory ByteClientFactory,
	adapters map[string]RPCAdapter,
) core.Factory {
	return func(config core.BusConfig) (core.MessageBus, error) {
		ddsConfig, typeName, qos, err := ConfigFromOptions(config.Options)
		if err != nil {
			return nil, err
		}
		if config.Name != "" {
			ddsConfig.ParticipantName = config.Name
		}
		client, err := factory(ddsConfig)
		if err != nil {
			return nil, err
		}
		return NewBusWithKindAndRPCAdapters(kind, ddsConfig, typeName, qos, client, adapters), nil
	}
}

func Register(factory ByteClientFactory) {
	core.Register(communication.TransportCycloneDDS, NewBusFactory(factory))
}

func RegisterWithRPCAdapters(factory ByteClientFactory, adapters map[string]RPCAdapter) {
	core.Register(communication.TransportCycloneDDS, NewBusFactoryWithRPCAdapters(factory, adapters))
}

func RegisterWithKindAndRPCAdapters(kind communication.TransportKind, factory ByteClientFactory, adapters map[string]RPCAdapter) {
	core.Register(kind, NewBusFactoryWithKindAndRPCAdapters(kind, factory, adapters))
}

func (b *Bus) Kind() communication.TransportKind {
	if b.kind == "" {
		return communication.TransportCycloneDDS
	}
	return b.kind
}

func (b *Bus) Capabilities() core.Capabilities {
	return core.Capabilities{PublishSubscribe: true, RequestReply: true}
}

func (b *Bus) Connect(ctx context.Context) error {
	return b.client.Connect(ctx, b.config)
}

func (b *Bus) Close(ctx context.Context) error {
	for _, adapter := range b.adapters {
		if closer, ok := adapter.(interface {
			CloseClient(ByteClient)
		}); ok {
			closer.CloseClient(b.client)
		}
	}
	return b.client.Close(ctx)
}

func (b *Bus) Publish(ctx context.Context, channel core.Channel, payload []byte) error {
	topic := b.topic(channel)
	if err := b.ensurePublishReady(ctx, topic); err != nil {
		return err
	}
	return b.client.Publish(ctx, topic, payload)
}

func (b *Bus) Subscribe(ctx context.Context, channel core.Channel, handler core.BytesHandler) error {
	return b.client.Subscribe(ctx, Subscription{Topic: b.topic(channel)}, ByteHandler(handler))
}

func (b *Bus) Request(
	ctx context.Context,
	channel core.Channel,
	payload []byte,
	timeout time.Duration,
) ([]byte, error) {
	binding := b.rpcBinding(channel)
	adapter := b.adapters[binding.Standard]
	if adapter == nil {
		return nil, fmt.Errorf(
			"cyclonedds request/reply requires infra DDS RPC adapter for %s",
			binding.Standard,
		)
	}
	if clientAdapter, ok := adapter.(ClientRPCAdapter); ok {
		return clientAdapter.RequestWithClient(ctx, b.client, binding, payload, timeout)
	}
	return adapter.Request(ctx, binding, payload, timeout)
}

func (b *Bus) HandleRequest(ctx context.Context, channel core.Channel, handler core.RequestHandler) error {
	binding := b.rpcBinding(channel)
	adapter := b.adapters[binding.Standard]
	if adapter == nil {
		return fmt.Errorf(
			"cyclonedds request/reply requires infra DDS RPC adapter for %s",
			binding.Standard,
		)
	}
	if clientAdapter, ok := adapter.(ClientRPCAdapter); ok {
		return clientAdapter.HandleRequestWithClient(ctx, b.client, binding, handler)
	}
	return adapter.HandleRequest(ctx, binding, handler)
}

func (b *Bus) topic(channel core.Channel) TopicConfig {
	typeName := b.typeName
	if typed := typedDDSType(channel); typed != "" {
		if supporter, ok := b.client.(interface {
			SupportsTypedDDS(string) bool
		}); ok && supporter.SupportsTypedDDS(typed) {
			typeName = typed
		}
	}
	return TopicConfig{
		TopicName: channel.Name,
		TypeName:  typeName,
		QoS:       mergeQoS(b.qos, channel.Metadata),
	}
}

func (b *Bus) ensurePublishReady(ctx context.Context, topic TopicConfig) error {
	if b.kind != communication.TransportFastDDS {
		return nil
	}
	key := topicRPCTopicKey(topic)
	b.publishReadyMu.Lock()
	if b.publishReady[key] {
		b.publishReadyMu.Unlock()
		return nil
	}
	b.publishReadyMu.Unlock()

	if err := b.client.PreparePublish(ctx, topic); err != nil {
		return err
	}
	timeout := fastDDSPublishMatchTimeout()
	matched := true
	if timeout > 0 {
		waitCtx, cancel := context.WithTimeout(ctx, timeout)
		err := waitForSubscribers(waitCtx, b.client, topic)
		cancel()
		matched = err == nil
	}
	if matched {
		b.publishReadyMu.Lock()
		b.publishReady[key] = true
		b.publishReadyMu.Unlock()
	}
	return nil
}

func fastDDSPublishMatchTimeout() time.Duration {
	for _, key := range []string{"PR_FASTDDS_MATCH_TIMEOUT_SEC", "PR_MATRIX_DISCOVERY_WAIT_SEC"} {
		value := os.Getenv(key)
		if value == "" {
			continue
		}
		seconds, err := strconv.ParseFloat(value, 64)
		if err != nil {
			continue
		}
		if seconds <= 0 {
			return 0
		}
		return time.Duration(seconds * float64(time.Second))
	}
	return 500 * time.Millisecond
}

func (b *Bus) rpcBinding(channel core.Channel) RPCBinding {
	standard := normalizeStandard(channel.Metadata["rpc.standard"])
	requestName := channel.Metadata["rpc.request_channel"]
	if requestName == "" {
		requestName = channel.Name
	}
	responseName := channel.Metadata["rpc.response_channel"]
	if responseName == "" {
		responseName = requestName + ".reply"
	}
	qos := mergeQoS(b.qos, channel.Metadata)
	typeName := b.topic(channel).TypeName
	return RPCBinding{
		Standard: standard,
		RequestChannel: TopicConfig{
			TopicName: requestName,
			TypeName:  typeName,
			QoS:       qos,
		},
		ResponseChannel: TopicConfig{
			TopicName: responseName,
			TypeName:  typeName,
			QoS:       qos,
		},
	}
}

func typedDDSType(channel core.Channel) string {
	mode := strings.ReplaceAll(strings.ToLower(strings.TrimSpace(channel.Metadata["dds.mode"])), "-", "_")
	if mode != "typed" && mode != "typed_preferred" && strings.TrimSpace(channel.Metadata["schema.language"]) != "omg_idl" {
		return ""
	}
	if typ := strings.TrimSpace(channel.Metadata["dds.type"]); typ != "" {
		return typ
	}
	return strings.TrimSpace(channel.Metadata["schema.type"])
}

func normalizeAdapters(adapters map[string]RPCAdapter) map[string]RPCAdapter {
	out := map[string]RPCAdapter{}
	for key, adapter := range adapters {
		if adapter != nil {
			out[normalizeStandard(key)] = adapter
		}
	}
	return out
}

func normalizeStandard(value string) string {
	normalized := strings.ReplaceAll(strings.ToLower(strings.TrimSpace(value)), "-", "_")
	switch normalized {
	case "", "omg", "dds_rpc", "omg_dds_rpc":
		return "omg_dds_rpc"
	case "rmw", "rmw_cyclonedds", "rmw_cyclonedds_cpp", "ros2_rmw":
		return "rmw_cyclonedds"
	default:
		return normalized
	}
}

func ConfigFromOptions(options map[string]any) (CycloneDDSConfig, string, map[string]string, error) {
	config := DefaultConfig()
	config.DomainID = nativeDDSDomainID(options)
	typeName := "PacificRimMessageEnvelope"
	qos := map[string]string{}
	for key, value := range options {
		switch key {
		case "domain_id", "ros_domain_id", "native_domain_id", "native_domain_offset":
		case "participant_name", "name":
			config.ParticipantName = fmt.Sprint(value)
		case "config_uri":
			config.ConfigURI = fmt.Sprint(value)
		case "read_period_sec":
			config.ReadPeriodSec = toFloat(value, config.ReadPeriodSec)
		case "type_name":
			typeName = fmt.Sprint(value)
		case "qos":
			qos["profile"] = fmt.Sprint(value)
		default:
			if len(key) > 4 && key[:4] == "qos." {
				qos[key[4:]] = fmt.Sprint(value)
			}
		}
	}
	return config, typeName, qos, nil
}

func nativeDDSDomainID(options map[string]any) int {
	for _, key := range []string{"native_domain_id", "domain_id"} {
		if value, ok := options[key]; ok {
			return toInt(value, 0)
		}
	}
	if value := strings.TrimSpace(os.Getenv("PACIFIC_RIM_NATIVE_DDS_DOMAIN_ID")); value != "" {
		return toInt(value, 0)
	}
	base := 0
	if value, ok := options["ros_domain_id"]; ok {
		base = toInt(value, 0)
	} else if envDomainID := strings.TrimSpace(os.Getenv("ROS_DOMAIN_ID")); envDomainID != "" {
		base = toInt(envDomainID, 0)
	}
	return base + nativeDDSDomainOffset(options)
}

func nativeDDSDomainOffset(options map[string]any) int {
	if value, ok := options["native_domain_offset"]; ok {
		return toInt(value, 100)
	}
	if value := strings.TrimSpace(os.Getenv("PACIFIC_RIM_NATIVE_DDS_DOMAIN_OFFSET")); value != "" {
		return toInt(value, 100)
	}
	return 100
}

func mergeQoS(base map[string]string, metadata map[string]string) map[string]string {
	out := map[string]string{}
	for key, value := range base {
		out[key] = value
	}
	for key, value := range metadata {
		if key == "qos" {
			out["profile"] = value
		} else if len(key) > 4 && key[:4] == "qos." {
			out[key[4:]] = value
		}
	}
	return out
}

func toInt(value any, fallback int) int {
	switch typed := value.(type) {
	case int:
		return typed
	case int64:
		return int(typed)
	case float64:
		return int(typed)
	case string:
		parsed, err := strconv.Atoi(typed)
		if err == nil {
			return parsed
		}
	}
	return fallback
}

func toFloat(value any, fallback float64) float64 {
	switch typed := value.(type) {
	case float64:
		return typed
	case float32:
		return float64(typed)
	case int:
		return float64(typed)
	case int64:
		return float64(typed)
	case string:
		parsed, err := strconv.ParseFloat(typed, 64)
		if err == nil {
			return parsed
		}
	}
	return fallback
}

File: infra/communication/python/pacific_rim_communication_infra/dds/bus.py
from __future__ import annotations

import asyncio
import os
import time
from dataclasses import replace
from typing import Any, Mapping

from pacific_rim_communication_infra.contracts import TransportKind
from pacific_rim_communication_infra.core import (
  BytesHandler,
  ChannelLike,
  MiddlewareCapabilities,
  RequestHandler,
  channel_name,
)

from .cyclonedds import (
  CycloneDdsClient,
  CycloneDdsConfig,
  CycloneDdsRPCAdapter,
  CycloneDdsRPCBinding,
  CycloneDdsSubscription,
  CycloneDdsTopicConfig,
)


class CycloneDdsMessageBus:
  kind = TransportKind.CYCLONE_DDS
  capabilities = MiddlewareCapabilities(publish_subscribe=True, request_reply=True)

  def __init__(
    self,
    client: CycloneDdsClient,
    type_name: str = "PacificRimMessageEnvelope",
    rpc_adapters: Mapping[str, CycloneDdsRPCAdapter] | None = None,
  ):
    self._client = client
    self._type_name = type_name
    self._qos = dict(getattr(client.config, "qos", {}) or {})
    default_adapter = TopicRPCAdapter(client)
    self._rpc_adapters = {
      "omg_dds_rpc": default_adapter,
      "rmw_cyclonedds": default_adapter,
    }
    self._rpc_adapters.update({str(key): value for key, value in dict(rpc_adapters or {}).items()})
    self._default_adapter = default_adapter

  @classmethod
  def from_options(cls, options: Mapping[str, Any]) -> "CycloneDdsMessageBus":
    config = CycloneDdsConfig()
    values: dict[str, Any] = {"domain_id": native_domain_id_from_options(options)}
    for key in ("participant_name", "config_uri", "read_period_sec"):
      if key in options:
        values[key] = options[key]
    qos = {
      str(key)[4:]: value
      for key, value in options.items()
      if str(key).startswith("qos.")
    }
    if "qos" in options:
      qos["profile"] = options["qos"]
    if qos:
      values["qos"] = qos
    type_name = str(options.get("type_name", "PacificRimMessageEnvelope"))
    return cls(CycloneDdsClient(replace(config, **values)), type_name=type_name)

  async def connect(self) -> None:
    await self._client.connect()

  async def close(self) -> None:
    for adapter in {self._default_adapter, *self._rpc_adapters.values()}:
      close = getattr(adapter, "close", None)
      if close is not None:
        result = close()
        if asyncio.iscoroutine(result):
          await result
    await self._client.close()

  async def publish_bytes(self, channel: ChannelLike, payload: bytes) -> None:
    await self._client.publish(self._topic(channel), payload)

  async def subscribe_bytes(self, channel: ChannelLike, handler: BytesHandler) -> None:
    await self._client.subscribe(CycloneDdsSubscription(self._topic(channel)), handler)

  async def request_bytes(
    self,
    channel: ChannelLike,
    payload: bytes,
    timeout_sec: float = 2.0,
  ) -> bytes:
    binding = self._rpc_binding(channel)
    adapter = self._rpc_adapters.get(binding.standard)
    if adapter is None:
      raise NotImplementedError(
        "CycloneDDS request/reply requires an infra DDS RPC adapter for "
        f"{binding.standard}. Configure standard: omg_dds_rpc or "
        "standard: rmw_cyclonedds, and register the matching adapter in "
        "infra/communication/dds."
      )
    return await adapter.request(binding, payload, timeout_sec)

  async def handle_request_bytes(
    self,
    channel: ChannelLike,
    handler: RequestHandler,
  ) -> None:
    binding = self._rpc_binding(channel)
    adapter = self._rpc_adapters.get(binding.standard)
    if adapter is None:
      raise NotImplementedError(
        "CycloneDDS request/reply requires an infra DDS RPC adapter for "
        f"{binding.standard}. Configure standard: omg_dds_rpc or "
        "standard: rmw_cyclonedds, and register the matching adapter in "
        "infra/communication/dds."
      )
    await adapter.handle_request(binding, handler)

  def _topic(self, channel: ChannelLike) -> CycloneDdsTopicConfig:
    qos = dict(self._qos)
    metadata = channel.metadata if hasattr(channel, "metadata") else {}
    for key, value in dict(metadata or {}).items():
      if key == "qos":
        qos["profile"] = value
      elif str(key).startswith("qos."):
        qos[str(key)[4:]] = value
    type_name = self._type_name
    typed_type = _typed_dds_type(dict(metadata or {}))
    supports_typed = getattr(self._client, "supports_typed_dds", None)
    if typed_type and supports_typed is not None and supports_typed(typed_type):
      type_name = typed_type
    return CycloneDdsTopicConfig(topic_name=channel_name(channel), type_name=type_name, qos=qos)

  def _rpc_binding(self, channel: ChannelLike) -> CycloneDdsRPCBinding:
    metadata = channel.metadata if hasattr(channel, "metadata") else {}
    standard = self._standard(dict(metadata or {}).get("rpc.standard", ""))
    request_name = str(dict(metadata or {}).get("rpc.request_channel") or channel_name(channel)).strip()
    response_name = str(dict(metadata or {}).get("rpc.response_channel") or f"{request_name}.reply").strip()
    request_channel = CycloneDdsTopicConfig(
      topic_name=request_name,
      type_name=self._topic(channel).type_name,
      qos=self._qos_from_metadata(metadata),
    )
    response_channel = CycloneDdsTopicConfig(
      topic_name=response_name,
      type_name=self._topic(channel).type_name,
      qos=self._qos_from_metadata(metadata),
    )
    return CycloneDdsRPCBinding(
      standard=standard,
      request_channel=request_channel,
      response_channel=response_channel,
    )

  def _qos_from_metadata(self, metadata: Mapping[str, Any]) -> dict[str, Any]:
    qos = dict(self._qos)
    for key, value in dict(metadata or {}).items():
      if key == "qos":
        qos["profile"] = value
      elif str(key).startswith("qos."):
        qos[str(key)[4:]] = value
    return qos

  @staticmethod
  def _standard(value: Any) -> str:
    normalized = str(value or "").strip().lower().replace("-", "_")
    aliases = {
      "": "omg_dds_rpc",
      "omg": "omg_dds_rpc",
      "dds_rpc": "omg_dds_rpc",
      "omg_dds_rpc": "omg_dds_rpc",
      "rmw": "rmw_cyclonedds",
      "rmw_cyclonedds": "rmw_cyclonedds",
      "rmw_cyclonedds_cpp": "rmw_cyclonedds",
      "ros2_rmw": "rmw_cyclonedds",
    }
    return aliases.get(normalized, normalized)


class TopicRPCAdapter(CycloneDdsRPCAdapter):
  standard = "omg_dds_rpc"
  ready_period_sec = 0.005
  ready_wait_sec = 1.5
  probe_wait_sec = 1.5

  def __init__(self, client: CycloneDdsClient):
    self._client = client
    self._ready_tasks: set[asyncio.Task[None]] = set()

  async def request(
    self,
    binding: CycloneDdsRPCBinding,
    payload: bytes,
    timeout_sec: float,
  ) -> bytes:
    if timeout_sec <= 0:
      timeout_sec = 2.0
    loop = asyncio.get_running_loop()
    response_future: asyncio.Future[bytes] = loop.create_future()
    handshake_future: asyncio.Future[None] = loop.create_future()
    probe_frame = _new_rpc_probe_frame()
    ack_frame = _rpc_ack_frame_for_probe(probe_frame)

    async def on_response(data: bytes) -> None:
      if not response_future.done():
        response_future.set_result(bytes(data))

    unsubscribe_response = await self._client.subscribe(CycloneDdsSubscription(binding.response_channel), on_response)
    unsubscribe_ack = await self._client.subscribe(
      CycloneDdsSubscription(_rpc_probe_ack_channel(binding)),
      lambda data: _set_handshake_if_ack(data, ack_frame, handshake_future),
    )
    try:
      await self._client.prepare_publish(binding.request_channel)
      await self._client.prepare_publish(_rpc_probe_channel(binding))
      if await self._wait_for_ready(binding, min(timeout_sec, self.ready_wait_sec)):
        await self._wait_for_subscribers(binding.request_channel, min(timeout_sec, 1.0))
        await self._wait_for_subscribers(_rpc_probe_channel(binding), min(timeout_sec, 0.5))
        await self._probe_rpc_pairing(
          binding,
          probe_frame,
          handshake_future,
          min(timeout_sec, self.probe_wait_sec),
        )
      await self._client.publish(binding.request_channel, payload)
      return await asyncio.wait_for(response_future, timeout=timeout_sec)
    finally:
      unsubscribe_response()
      unsubscribe_ack()

  async def handle_request(
    self,
    binding: CycloneDdsRPCBinding,
    handler: RequestHandler,
  ) -> None:
    async def on_request(data: bytes) -> None:
      result = handler(bytes(data))
      if asyncio.iscoroutine(result):
        result = await result
      await self._wait_for_subscribers(binding.response_channel, 1.0)
      await self._client.publish(binding.response_channel, bytes(result or b""))

    await self._client.prepare_publish(binding.response_channel)
    await self._client.prepare_publish(_rpc_probe_ack_channel(binding))
    await self._client.subscribe(CycloneDdsSubscription(binding.request_channel), on_request)
    await self._client.subscribe(
      CycloneDdsSubscription(_rpc_probe_channel(binding)),
      lambda data: self._handle_probe(binding, data),
    )
    await self._start_ready_publisher(binding)

  async def close(self) -> None:
    for task in list(self._ready_tasks):
      task.cancel()
    if self._ready_tasks:
      await asyncio.gather(*self._ready_tasks, return_exceptions=True)
    self._ready_tasks.clear()

  async def _wait_for_ready(self, binding: CycloneDdsRPCBinding, timeout_sec: float) -> bool:
    loop = asyncio.get_running_loop()
    ready: asyncio.Future[None] = loop.create_future()

    async def on_ready(_: bytes) -> None:
      if not ready.done():
        ready.set_result(None)

    unsubscribe = await self._client.subscribe(CycloneDdsSubscription(_rpc_ready_channel(binding)), on_ready)
    try:
      await asyncio.wait_for(ready, timeout=max(0.0, timeout_sec))
      return True
    except asyncio.TimeoutError:
      return False
    finally:
      unsubscribe()

  async def _wait_for_subscribers(self, topic: CycloneDdsTopicConfig, timeout_sec: float) -> bool:
    wait_for_subscribers = getattr(self._client, "wait_for_subscribers", None)
    if wait_for_subscribers is None:
      return True
    waited = wait_for_subscribers(topic, timeout_sec=timeout_sec)
    if asyncio.iscoroutine(waited):
      return bool(await waited)
    return bool(waited)

  async def _wait_for_publishers(self, topic: CycloneDdsTopicConfig, timeout_sec: float) -> bool:
    wait_for_publishers = getattr(self._client, "wait_for_publishers", None)
    if wait_for_publishers is None:
      return True
    waited = wait_for_publishers(topic, timeout_sec=timeout_sec)
    if asyncio.iscoroutine(waited):
      return bool(await waited)
    return bool(waited)

  async def _probe_rpc_pairing(
    self,
    binding: CycloneDdsRPCBinding,
    probe_frame: bytes,
    handshake_future: asyncio.Future[None],
    timeout_sec: float,
  ) -> None:
    deadline = asyncio.get_running_loop().time() + max(0.0, timeout_sec)
    while not handshake_future.done() and asyncio.get_running_loop().time() < deadline:
      await self._client.publish(_rpc_probe_channel(binding), probe_frame)
      try:
        await asyncio.wait_for(
          asyncio.shield(handshake_future),
          timeout=min(0.005, max(0.0, deadline - asyncio.get_running_loop().time())),
        )
      except asyncio.TimeoutError:
        pass
    if not handshake_future.done():
      raise TimeoutError(f"CycloneDDS RPC endpoint pairing timed out: {binding.request_channel.topic_name}")

  async def _start_ready_publisher(self, binding: CycloneDdsRPCBinding) -> None:
    channel = _rpc_ready_channel(binding)
    await self._client.prepare_publish(channel)

    async def publish_ready() -> None:
      while self._client.is_connected:
        await self._wait_for_publishers(binding.request_channel, 0.1)
        await self._wait_for_subscribers(channel, 0.1)
        await self._client.publish(channel, b"ready")
        await asyncio.sleep(self.ready_period_sec)

    task = asyncio.create_task(publish_ready())
    self._ready_tasks.add(task)
    task.add_done_callback(self._ready_tasks.discard)

  async def _handle_probe(self, binding: CycloneDdsRPCBinding, data: bytes) -> None:
    if not data.startswith(_RPC_PROBE_PREFIX):
      return
    ack_channel = _rpc_probe_ack_channel(binding)
    await self._wait_for_subscribers(ack_channel, 0.5)
    await self._client.publish(ack_channel, _rpc_ack_frame_for_probe(data))


def _rpc_ready_channel(binding: CycloneDdsRPCBinding) -> CycloneDdsTopicConfig:
  return replace(binding.request_channel, topic_name=f"{binding.request_channel.topic_name}.__pr_ready")


def _rpc_probe_channel(binding: CycloneDdsRPCBinding) -> CycloneDdsTopicConfig:
  return replace(binding.request_channel, topic_name=f"{binding.request_channel.topic_name}.__pr_probe")


def _rpc_probe_ack_channel(binding: CycloneDdsRPCBinding) -> CycloneDdsTopicConfig:
  return replace(binding.response_channel, topic_name=f"{binding.response_channel.topic_name}.__pr_probe_ack")


def _typed_dds_type(metadata: Mapping[str, Any]) -> str:
  mode = str(metadata.get("dds.mode") or "").strip().lower().replace("-", "_")
  language = str(metadata.get("schema.language") or "").strip().lower()
  if mode not in {"typed", "typed_preferred"} and language != "omg_idl":
    return ""
  return str(metadata.get("dds.type") or metadata.get("schema.type") or "").strip()


_RPC_PROBE_PREFIX = b"\x00PRPC_READY_V1\x00probe:"
_RPC_ACK_PREFIX = b"\x00PRPC_READY_V1\x00ack:"


def _new_rpc_probe_frame() -> bytes:
  return _RPC_PROBE_PREFIX + str(time.time_ns()).encode("ascii")


def _rpc_ack_frame_for_probe(probe: bytes) -> bytes:
  return _RPC_ACK_PREFIX + bytes(probe[len(_RPC_PROBE_PREFIX):])


def _set_handshake_if_ack(data: bytes, ack_frame: bytes, future: asyncio.Future[None]) -> None:
  if data == ack_frame and not future.done():
    future.set_result(None)


def _optional_int(value: Any) -> int | None:
  if value is None or value == "":
    return None
  return int(value)


def native_domain_id_from_options(options: Mapping[str, Any]) -> int:
  for key in ("native_domain_id", "domain_id"):
    value = _optional_int(options.get(key))
    if value is not None:
      return value
  native_env = _optional_int(os.environ.get("PACIFIC_RIM_NATIVE_DDS_DOMAIN_ID"))
  if native_env is not None:
    return native_env
  ros_domain = _optional_int(options.get("ros_domain_id"))
  if ros_domain is None:
    ros_domain = _optional_int(os.environ.get("ROS_DOMAIN_ID")) or 0
  return ros_domain + native_domain_offset_from_options(options)


def native_domain_offset_from_options(options: Mapping[str, Any]) -> int:
  offset = _optional_int(options.get("native_domain_offset"))
  if offset is not None:
    return offset
  env_offset = _optional_int(os.environ.get("PACIFIC_RIM_NATIVE_DDS_DOMAIN_OFFSET"))
  return env_offset if env_offset is not None else 100
